src/QsCompiler/SyntaxProcessor/StatementVerification.fs

```diff
@@ -133,77 +97,45 @@ let NewReturnStatement comments (location: QsLocation) (context: ScopeContext) e
-                    | _ -> Seq.zip syms ts |> Seq.map (fun (s, t) -> GetBindings(s, t, None))
-                    |> Seq.toList
-                    |> List.unzip3
-
-                symbolTuple symItems, Array.concat declarations, Array.concat errs
-            | _ ->
-                let (symItems, declarations, errs) =
-                    syms
-                    |> Seq.map (fun s -> (s, InvalidType |> ResolvedType.New, None) |> GetBindings)
-                    |> Seq.toList
-                    |> List.unzip3
-
-                let errs =
-                    if exType.isInvalid then
-                        errs
-                    else
-                        [|
-                            sym.RangeOrDefault
-                            |> QsCompilerDiagnostic.Error(ErrorCode.SymbolTupleShapeMismatch, [ exType |> toString ])
-                        |]
-                        :: errs
-
-                symbolTuple symItems, Array.concat declarations, Array.concat errs
-
-    GetBindings(qsSym, rhsType, rhsEx)
+let rec private VerifyBinding (inference: InferenceContext) tryBuildDeclaration (symbol, rhsType) warnOnDiscard =
+    let symbolTuple =
+        function
+        | [] -> failwith "Symbol tuple is empty."
+        | [ x ] -> x
+        | xs -> ImmutableArray.CreateRange xs |> VariableNameTuple
+
+    match symbol.Symbol with
+    | InvalidSymbol -> InvalidItem, [||], [||]
+    | MissingSymbol when warnOnDiscard ->
+        let warning = QsCompilerDiagnostic.Warning (WarningCode.DiscardingItemInAssignment, []) symbol.RangeOrDefault
+        DiscardedItem, [||], [| warning |]
+    | MissingSymbol -> DiscardedItem, [||], [||]
+    | OmittedSymbols
+    | QualifiedSymbol _ ->
+        let error = QsCompilerDiagnostic.Error (ErrorCode.ExpectingUnqualifiedSymbol, []) symbol.RangeOrDefault
+        InvalidItem, [||], [| error |]
+    | Symbol name ->
+        match tryBuildDeclaration (name, symbol.RangeOrDefault) (rhsType) with
+        | Some declaration, diagnostics -> VariableName name, [| declaration |], diagnostics
+        | None, diagnostics -> VariableName name, [||], diagnostics
+    | SymbolTuple symbols ->
+        let types = symbols |> Seq.map (fun symbol -> inference.Fresh symbol.RangeOrDefault) |> Seq.toList
+
+        let tupleType =
+            ImmutableArray.CreateRange types
+            |> TupleType
+            |> ResolvedType.create (TypeRange.inferred symbol.Range)
+
+        let unifyDiagnostics = inference.Unify(tupleType, rhsType)
+
+        let verify symbol symbolType =
+            VerifyBinding inference tryBuildDeclaration (symbol, symbolType) warnOnDiscard
+
+        let combine (item, declarations1, diagnostics1) (items, declarations2, diagnostics2) =
+            item :: items, Array.append declarations1 declarations2, Array.append diagnostics1 diagnostics2
+
+        let items, declarations, diagnostics = Seq.foldBack combine (Seq.map2 verify symbols types) ([], [||], [||])
```

Comment:
What if the symbols and types don't have the same length?

Reply:
That is not possible, because `types` is a mapped version of `symbols`.

```fsharp
let types = symbols |> Seq.map (fun symbol -> inference.Fresh symbol.RangeOrDefault) |> Seq.toList
```

Reply:
What I am looking for is where the check went that the symbol tuple shape is compatible with the expression types on the right hand side.

Reply:
That happens in unification:

https://github.com/microsoft/qsharp-compiler/blob/b2bdd30cc0835fdae9e697504af48363439cbd1e/src/QsCompiler/SyntaxProcessor/StatementVerification.fs#L129

`tupleType` is the type of the symbol tuple containing fresh type parameters `('t1, 't2, ..., 'tn)`. If the type of the right-hand side has a different tuple shape, the types will not unify.